[SAML NewDot] Add SAML flow for web, mweb, desktop

NikkiWines opened the pull request:
### Details
Adds support for signing in with SAML on web and mweb.

### Fixed Issues
$ https://github.com/Expensify/App/issues/28215
$ https://github.com/Expensify/App/issues/28216
PROPOSAL: N/A Internal 

### Tests
(Internal) 
(for Web, mWeb, desktop)
1. Follow steps in https://github.com/Expensify/Web-Expensify/pull/38972 to get set up with SAML locally
2. Configure your account to have SAML enabled but not required in oldDot 
3. Enter the email for your account in newDot 
4. Confirm you see the option to choose to sign in with SAML or with a magic code 
5. Choose the `Go Back` option and confirm you're dropped back on the sign-in page with your email pre-filled
6. Press `Continue`
7. Choose the `Use Magic Code` option and confirm you get a magic code sent to your email 
8. Press `Go Back` and confirm you're back on the sign-in page with your email pre-filled
9. Press `Continue`
10. Choose `Use Single Sign On` option
11. Confirm you briefly see an interstitial page about launching your SSO provider's login portal
12. _(web/mweb only)_ Navigate back to the root page (i.e. localhost:8082/ or new.expensify.com) 
13. _(web/mweb only)_  Confirm you land on the page with the option to choose to sign in with SAML or with a magic code 
14. _(web/mweb only)_  Choose `Use Single Sign On` option
15. Enter your SSO credentials 
16. Confirm you're redirected back to New Expensify and are logged into your account 
17. In oldDot, set SAML to required 
18. In a new incognito window, open New Expensify and enter your email address
19. Confirm you're automatically redirected to your SSO providers login page (with the interstitial page being briefly shown) 
12. _(web/mweb only)_  Navigate back to the root page (i.e. localhost:8082/ or new.expensify.com) 
13. _(web/mweb only)_  Confirm you land on the sign-in page _without_ your login prefilled
14. _(web/mweb only)_  Enter the email for a non-SAML account and confirm you're directed to the magic code view for that login
15. _(web/mweb only)_  Click go back and enter your SAML required login once again.
16. _(web/mweb only)_  Confirm you're automatically redirected to your SSO providers login page
21. Enter your credentials and confirm you're successfully logged into New Expensify 

(for ios/android) 
1. Log into the same account you used in step 17 above 
2. Confirm you see an error showing that you cannot log into an account with SAML required 

(External) 
(for Web, mWeb, Desktop) 
1. Create an account with a validated domain on expensify.com
2. Configure your account to have SAML enabled but not required
3. Enter the email for your account in New Expensify 
4. Confirm you see the option to choose to sign in with SAML or with a magic code 
5. Choose the `Go Back` option and confirm you're dropped back on the sign-in page with your email pre-filled
6. Press `Continue`
7. Choose the `Use Magic Code` option and confirm you get a magic code sent to your email 
8. Press `Go Back` and confirm you're back on the sign-in page with your email pre-filled
9. Press `Continue`
10. Choose `Use Single Sign On` option
11. Confirm you briefly see an interstitial page about launching your SSO provider's login portal
12. _(web/mweb only)_ Navigate back to the root page (i.e. localhost:8082/ or new.expensify.com) 
13. _(web/mweb only)_  Confirm you land on the page with the option to choose to sign in with SAML or with a magic code 
14. _(web/mweb only)_  Choose `Use Single Sign On` option
15. In expensify.com, set SAML to required 
16. In a new incognito window, open New Expensify and enter your email address
17. Confirm you're automatically redirected to your SSO providers login page (with the interstitial page being briefly shown) 
18. _(web/mweb only)_  Navigate back to the root page (i.e. localhost:8082/ or new.expensify.com) 
19. _(web/mweb only)_  Confirm you land on the sign-in page _without_ your login prefilled
20. _(web/mweb only)_  Enter the email for a non-SAML account and confirm you're directed to the magic code view for that login
21. _(web/mweb only)_  Click go back and enter your SAML required login once again.
22. _(web/mweb only)_  Confirm you're automatically redirected to your SSO providers login page

(for ios/android) 
1. Log into the same account you used in step 16 above 
2. Confirm you see an error showing that you cannot log into an account with SAML required 

### Offline tests
1. Navigate to the login page for New Expensify and enter your email 
2. Confirm you see the 
3. Disable your wifi or disable your network connections 
4. Confirm you see the offline indicator

<img width="418" alt="Screenshot 2023-09-28 at 15 47 43" src="https://github.com/Expensify/App/assets/16747822/d1aab1ca-c3e3-4a7d-8f17-81a4152e684b">

### QA Steps
Same as test steps 

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

**SAML Required Flow**

https://github.com/Expensify/App/assets/16747822/b7327bbd-f9e7-4cef-9e62-34e0f1cfd6bd

**SAML Enabled Flow**

https://github.com/Expensify/App/assets/16747822/f5694691-9638-4ed3-a013-0428c28c4788

</details>

<details>
<summary>Mobile Web - Chrome</summary>

**SAML Required Flow**

**SAML Enabled Flow**

https://github.com/Expensify/App/assets/16747822/a87846df-1cb6-420e-a9e0-e3a72f4cf502

</details>

<details>
<summary>Mobile Web - Safari</summary>

**SAML Required Flow**

https://github.com/Expensify/App/assets/16747822/da60c9a1-7331-4afe-bc77-386c8801e2b8

**SAML Enabled Flow**

https://github.com/Expensify/App/assets/16747822/19e822eb-cda7-41d3-9219-1112e6bfce4f

</details>

<details>
<summary>Desktop</summary>

**SAML Required Flow** 

https://github.com/Expensify/App/assets/16747822/c37d7315-9a31-4b41-8dc8-cff8fe660cfc

**SAML Enabled Flow**

https://github.com/Expensify/App/assets/16747822/5e5f5961-6d69-40ff-a397-753c0fd2fc9f

</details>

<details>
<summary>iOS</summary>

**SAML Required Flow**

https://github.com/Expensify/App/assets/16747822/1ab1ee61-3b6e-4b3f-80b5-7341bd76b9b7

(SAML Enabled is the same as the current magic code flow) 
</details>

<details>
<summary>Android</summary>

**SAML Required Flow**

(SAML Enabled is the same as the current magic code flow) 
</details>

NikkiWines committed:
add SAMLEnabledForm

NikkiWines committed:
update SignInPage to route to SAMLEnabledForm and SAMLSignInPage

NikkiWines committed:
add SAMLSignInPage and routing

NikkiWines committed:
add beta and consts

NikkiWines committed:
add routing for SAML required flow

NikkiWines committed:
Merge branch 'main' of github.com:Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
add updated copy for SAML sign in page

NikkiWines committed:
redirect back to SignInPage instead of showing ValidateCodePage in subview

NikkiWines committed:
fix login redirecting when SAML enabled user opts to sign in with a magic code

NikkiWines committed:
add saml spanish translations

NikkiWines committed:
minor style and props validation

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -59,23 +74,37 @@ const defaultProps = {
+
+    // SAML is temporarily restricted to users on the beta or to users signing in on web and mweb
+    if (Permissions.canUseSAML() || platform === CONST.PLATFORM.WEB) {
+        shouldShowSAMLEnabledForm = hasAccount && hasLogin && isSAMLEnabled && !isSAMLRequired && isUsingSAMLLogin;
```
Sorry, why ` !SAMLRequired` is here?
Shouldn't we display form if SAML is required?

MonilBhavsar replied:
Oh I see, looks like we don't initiate directly and give an option to choose if it is not required

NikkiWines replied:
Correct!

MonilBhavsar commented on src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js:
```diff
@@ -196,6 +205,8 @@ function BaseValidateCodeForm(props) {
      */
     const clearSignInData = () => {
+        // Reset the user's preference for signing in with SAML
+        props.setIsUsingSAMLLogin(props.account.isSAMLEnabled || props.account.isSAMLRequired)
```
Should we set false here, so when user enters new email, we set the state again?

NikkiWines replied:
Oh, yep - i think you're right and false would work here

NikkiWines replied:
Actually...  I think we want to retain this because if the user chooses the magic code one time we don't want to lock the into that option - the next time they try to log in they should be able to choose again whether they want SAML or the magic code

MonilBhavsar commented on src/pages/signin/SAMLEnabledForm.js:
```diff
@@ -0,0 +1,109 @@
+                        props.setIsUsingSAMLLogin(false);
+                    }}
+                />
+                <ChangeExpensifyLoginLink onPress={() => Session.clearSignInData()} />
```
Also, set `isUsingSAMLLogin` to false here?

NikkiWines replied:
We can, but in this case they haven't actually made a decision on which method to use, they're just re-starting the sign in process 🤔

MonilBhavsar replied:
Yes, so they will use a new login and `isUsingSAMLLogin` will be populated again?
I don't think we'll use the value if we don't clear it. So better to clear it?

NikkiWines replied:
Yeah, I mean we basically just want to clear the state whenever the `LoginPage` is shown but setting it to `false` here might not be correct (i.e. if the new login has SAML enabled as well)

MonilBhavsar commented on src/languages/es.ts:
```diff
@@ -353,6 +353,14 @@ export default {
+        useSingleSignOn: 'Utilizar el inicio de sesión único',
+        useMagicCode: 'Usar código mágico',
+        launching: 'Cargando...',
+        oneMoment:'Un momento mientras te redirigimos al portal de inicio de sesión único de tu empresa.',
```
Same here

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -59,23 +74,37 @@ const defaultProps = {
  * @returns {Object}
  */
-function getRenderOptions({hasLogin, hasValidateCode, hasAccount, isPrimaryLogin, isAccountValidated, hasEmailDeliveryFailure}) {
+function getRenderOptions({hasLogin, hasValidateCode, hasAccount, isPrimaryLogin, isAccountValidated, isSAMLEnabled, isSAMLRequired, isUsingSAMLLogin, hasEmailDeliveryFailure}) {
```
Should we add these params in JSDoc comment?

NikkiWines replied:
Yes 🫣

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -59,23 +74,37 @@ const defaultProps = {
+    // SAML is temporarily restricted to users on the beta or to users signing in on web and mweb
+    if (Permissions.canUseSAML() || platform === CONST.PLATFORM.WEB) {
+        shouldShowSAMLEnabledForm = hasAccount && hasLogin && isSAMLEnabled && !isSAMLRequired && isUsingSAMLLogin;
+        shouldInitiateSAMLLogin = hasAccount && hasLogin && isSAMLRequired;
```
NAB, should we add `&& isSAMLEnabled`?

NikkiWines replied:
No, because SAML if SAML is required it's also enabled 😅 the latter is the stricter condition of the two and we only want to initiate the SAML login process if they have it required (not if it's only enabled)

MonilBhavsar replied:
Makes sense, just a safe check that it is enabled and required

situchan commented:
Lint is failing

Is it ready for external review or wait until https://github.com/Expensify/Web-Expensify/pull/38972 is deployed?

MonilBhavsar commented:
I think this should be on HOLD until Web-E PR is deployed. But yes, It can be reviewed

NikkiWines committed:
include login in useEffect dependency array

NikkiWines commented on src/languages/en.ts:
```diff
@@ -362,6 +362,14 @@ export default {
+        welcomeSAMLEnabled: 'Continue logging in with your company\'s Single Sign-On:',
+        orContinueWithMagicCode: 'Or optionally, your company allows signing in with a magic code',
+        useSingleSignOn: 'Use Single Sign-On',
+        useMagicCode: 'Use Magic Code',
```
ooh, hmm - yes i think so. good catch

NikkiWines replied:
actually.. wait, no i think it would look weird then cause it would be the only bit that's lower case 🤔

sakluger replied:
Can we please make the following copy changes?

1. Change 'welcomeSAMLEnabled' message to 'Continue logging in with single sign-on:'
2. Change 'useSingleSignOn' button text to 'Use single sign-on'
3. Change 'useMagicCode' button text to 'Use magic code'

NikkiWines commented:
@situchan you can review the portions under the `(external)` section in the Tests for now but the actually redirect to the SAML provider and then back to New Expensify won't work until the internal PR is deployed to prod

NikkiWines committed:
minor style and linting fixes

sakluger commented on src/languages/es.ts:
Let's change the spanish 'welcomeSAMLEnabled' message too to remove "your company's".

sakluger commented on src/languages/en.ts:
```diff
@@ -362,6 +362,14 @@ export default {
+        useSingleSignOn: 'Use Single Sign-On',
+        useMagicCode: 'Use Magic Code',
+        launching: 'Launching...',
+        oneMoment: 'One moment while we redirect you to your company\'s Single Sign-On portal.',
```
We should make 'single sign-on' lower case.

sakluger commented on src/languages/es.ts:
```diff
@@ -353,6 +353,14 @@ export default {
         privacy: 'Privacidad',
     },
+    samlSignIn: {
+        welcomeSAMLEnabled: 'Continua iniciando sesión con el inicio de sesión único de su empresa:',
```
Let's change the spanish 'welcomeSAMLEnabled' message too to remove "your company's".

NikkiWines committed:
update copy

NikkiWines committed:
prettier style

situchan commented on src/pages/signin/SAMLEnabledForm.js:
```diff
@@ -0,0 +1,109 @@
+    account: {},
+};
+
+function SAMLEnabledForm(props) {
```
please use destructured props

situchan commented on src/pages/signin/SAMLEnabledForm.js:
```diff
@@ -0,0 +1,109 @@
+SAMLEnabledForm.displayName = 'SAMLEnabledForm';
+
+export default compose(
+    withLocalize,
+    withWindowDimensions,
+    withNetwork(),
```
use hooks - `useLocalize`, `useWindowDimensions`, `useNetwork`

situchan commented on src/pages/signin/SAMLSignInPage/index.js:
```diff
@@ -0,0 +1,68 @@
+    credentials: {},
+};
+
+function SAMLSignInPage(props) {
```
Same here. Use destructured props

situchan commented on src/pages/signin/SAMLSignInPage/index.js:
```diff
@@ -0,0 +1,68 @@
+SAMLSignInPage.defaultProps = defaultProps;
+
+export default compose(
+    withLocalize,
```
`useLocalize` hook

situchan commented on src/pages/signin/SAMLSignInPage/index.js:
```diff
@@ -0,0 +1,68 @@
+
+function SAMLSignInPage(props) {
+    useEffect(() => {
+        window.open(`${CONFIG.EXPENSIFY.SAML_URL}?email=${props.credentials.login}&referer=${CONFIG.EXPENSIFY.EXPENSIFY_CASH_REFERER}`, '_self');
```
It's possible that this page can be accessed on native when `newDotSAML` beta enabled

https://github.com/Expensify/App/blob/4308871ec475d6346406b30a1302765eb0d30793/src/pages/signin/SignInPage.js#L92-L96

NikkiWines replied:
🤔 why would it crash? `credentials.login` is the only portion we'd need to be defined and we'll only show the `SAMLEnabledForm` (now `ChooseSSOOrMagicCode`) or intiaite the SAML login (i.e. navigate to that URL immediately) if the login is set.

situchan reviewed (commented):
Am I missing any config? I am trying to follow External step but I am not able to reach SAML page on web. I enabled permission in Permissions.ts


https://github.com/Expensify/App/assets/108292595/52eff30a-95f3-4b15-ad7c-18969cb01b8c

NikkiWines commented:
ah yes, @situchan we'll need to add you to the beta. I've added situchandrashil7+2@gmail.com to the beta just in case, but that account cannot have SAML set up per the `external` instructions above because it is on a public domain. 

Let me know if you have any other emails you want to test with and I'll add them to the beta as necessary

Lastly, you could manually set `isSAMLEnabled` and `isSAMLRequired` to true/false (as necessary) to test this functionality locally

NikkiWines committed:
use destructured props and hooks

NikkiWines committed:
rename file according to conventions

NikkiWines commented:
Updated!

NikkiWines committed:
remove useless compose

NikkiWines committed:
minor cleanup

NikkiWines committed:
prettier

situchan commented:
> Lastly, you could manually set `isSAMLEnabled` and `isSAMLRequired` to true/false (as necessary) to test this functionality locally

If enabled them, can I test using public domain email?

NikkiWines commented:
@situchan no - a private domain is required as SAML is a domain level setting

NikkiWines commented:
Off hold now that the internal PR is on production

NikkiWines committed:
fix state logic for whether or not the user has opted to use magic codes instead of saml

NikkiWines commented:
@MonilBhavsar @situchan this should be ready for another round of reviews :bow:

NikkiWines commented on src/pages/signin/SignInPage.js:
```diff
@@ -91,21 +123,39 @@ function SignInPage({credentials, account, isInModal}) {
+        hasAccount: !_.isEmpty(account),
+        isPrimaryLogin: !account.primaryLogin || account.primaryLogin === credentials.login,
+        isAccountValidated: Boolean(account.validated),
+        isSAMLEnabled: Boolean(account.isSAMLEnabled),
+        isSAMLRequired: Boolean(account.isSAMLRequired),
```
@situchan manually setting these to true (or false, as necessary) will simulate the user having SAML set up on a private domain. If you are having difficulty getting that set up for testing, you can just edit these to go through the flow - though of course then the actual SSO login testing + callback will not work 😅

MonilBhavsar commented on src/SCREENS.ts:
```diff
@@ -24,4 +24,5 @@ export default {
     SIGN_IN_WITH_APPLE_DESKTOP: 'AppleSignInDesktop',
     SIGN_IN_WITH_GOOGLE_DESKTOP: 'GoogleSignInDesktop',
     DESKTOP_SIGN_IN_REDIRECT: 'DesktopSignInRedirect',
+    SAML_SIGN_IN: 'SAMLSignIN',
```
NAB, Use lower case in value?
```suggestion
    SAML_SIGN_IN: 'SAMLSignIn',
```

MonilBhavsar commented on src/pages/signin/ChooseSSOOrMagicCode.js:
```diff
@@ -0,0 +1,101 @@
+        isLoading: PropTypes.bool,
+
+        /** Form that is being loaded */
+        loadingForm: PropTypes.bool,
```
`loadingForm` should be type of `ValueOf<typeof CONST.FORMS>;`?

NikkiWines replied:
Hmm, that's for typed files I think but maybe `PropTypes.oneOf(CONST.FORMS)`?

NikkiWines replied:
Actually i think this should be of type string

MonilBhavsar replied:
Yes right, I think we can use - PropTypes.oneOf(_.values(CONST.FORMS))

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -37,6 +43,15 @@ const propTypes = {
         /** Is this account having trouble receiving emails */
         hasEmailDeliveryFailure: PropTypes.bool,
+
+        /** Whether the account is in a loading state */
```
I think this comment used above is better
`/** Whether or not a sign on form is loading (being submitted) */`

MonilBhavsar reviewed (commented):
Few comments else looks good!

NikkiWines committed:
minor pr comments

NikkiWines commented:
done!

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
add support for showing error messages when returned in url

NikkiWines committed:
add desktop support because it works out of the box ????

NikkiWines committed:
Revert "add desktop support because it works out of the box ????" - it was too good to be true

NikkiWines commented:
Updating some things briefly for the SAML required flow (which may also fix the desktop routing) - should be updated shortly!

NikkiWines committed:
some edge case handling and saml required redirect logic

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
don't re-call signInWithShortLivedAUthToken if we're already loading the account

NikkiWines committed:
fix props

NikkiWines committed:
some linter fixes

NikkiWines commented:
ok this _should_ be ready hopefully for a final review :bow:

NikkiWines committed:
minor style

MonilBhavsar commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -85,7 +94,15 @@ function LogInWithShortLivedAuthTokenPage(props) {
                 <View style={styles.mt2}>
                     <Text style={[styles.fontSizeNormal, styles.textAlignCenter]}>
-                        {props.translate('deeplinkWrapper.expired')} <TextLink onPress={() => Navigation.navigate()}>{props.translate('deeplinkWrapper.signIn')}</TextLink>
+                        {props.translate('deeplinkWrapper.expired')}{' '}
```
NAB, we can move {' '} to new line

NikkiWines replied:
This change was from running `npm run prettier`

MonilBhavsar commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -105,9 +122,4 @@ LogInWithShortLivedAuthTokenPage.propTypes = propTypes;
-        account: {key: ONYXKEYS.ACCOUNT},
-    }),
-)(LogInWithShortLivedAuthTokenPage);
+export default compose(withLocalize, withOnyx({account: {key: ONYXKEYS.ACCOUNT}}), withOnyx({session: {key: ONYXKEYS.SESSION}}))(LogInWithShortLivedAuthTokenPage);
```
NAB, we can keep in new line for good readability

NikkiWines replied:
These changes were all from running `npm run prettier`

MonilBhavsar commented on src/pages/signin/LoginForm/BaseLoginForm.js:
```diff
@@ -159,7 +159,7 @@ function LoginForm(props) {
         // Just call clearAccountMessages on the login page (home route), because when the user is in the transition route and not yet authenticated,
         // this component will also be mounted, resetting account.isLoading will cause the app to briefly display the session expiration page.
-        if (props.isFocused) {
+        if (props.isFocused && props.isVisible) {
```
Sorry, why this change?

NikkiWines replied:
This is so that if we get an error back in the URL after the user tries to sign in with SSO that we display the error message instead of clearing it. 

You can test this by navigating to http://localhost:8082/transition?error=meepmeep&exitTo=%2F

MonilBhavsar replied:
Clarification - the error is supposed to be displayed on ChooseSSOOrMagicCode or ValidateCodeForm page, right?
Simply navigating to the URL does not display an error on the BaseLoginForm.

NikkiWines replied:
Ah, yes - I was thinking on the ChooseSSOOrMagicCode view, though then it wouldn't show for SAML required users 🤔 

I think that can be handled more fully in a follow up PR though, I'd like to get this merged ASAP so it's easier to develop for the remaining platforms 🙏

MonilBhavsar replied:
Sounds good 👍

MonilBhavsar commented:
@situchan how is it going with testing? Is there anything we can help?

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
fix merge issue

NikkiWines commented:
@MonilBhavsar @situchan friendly bump here - would love to get this merged asap 🙇 🙏

NikkiWines committed:
prettier

MonilBhavsar commented:
I'll get to test this today!

MonilBhavsar commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

MonilBhavsar commented:
1. I see this console error when SAML required is true and user is automatically redirected to SSO login page

<img width="1728" alt="Screenshot 2023-10-09 at 5 46 54 PM" src="https://github.com/Expensify/App/assets/32012005/be919288-2a2e-4aab-8879-f7361f410f2d">


2. [CRITICAL] I can't seem to escape this blank screen. I set SAML required to true. Went to SSO sign in page. Press back button twice to land to http://localhost:8082/. I see the page below. I refreshed and I then see empty page. The workaround is to clear Onyx data.

<img width="1728" alt="Screenshot 2023-10-09 at 5 43 50 PM" src="https://github.com/Expensify/App/assets/32012005/430efe32-1ac0-4aad-b8ae-1e823b175135">


I'll test on mobile now...

NikkiWines commented:
> I see this console error when SAML required is true and user is automatically redirected to SSO login page

Looking into this now 

> [CRITICAL] I can't seem to escape this blank screen. I set SAML required to true. Went to SSO sign in page. Press back button twice to land to http://localhost:8082/. I see the page below. I refreshed and I then see empty page. The workaround is to clear Onyx data.

This is somewhat expected if the user has entered a login and they are required to use SAML to log in, because there's only one flow for them to go through to access their account. However, it's not an ideal experience in the event that they want to use another account. For OldDot it looks like we do something slightly different so maybe we can try and emulate that here 🤔 

https://github.com/Expensify/App/assets/16747822/1a51ad99-fe82-4762-9b2a-cb7031a7f0c2

However, I do think it's kind of a more dev-environement kind of flow. Most SAML required users should not have multiple accounts that they're trying to access. 

cc: @sakluger incase you have strong thoughts on this

sakluger commented:
> [CRITICAL] I can't seem to escape this blank screen. I set SAML required to true. Went to SSO sign in page. Press back button twice to land to http://localhost:8082/. I see the page below. I refreshed and I then see empty page. The workaround is to clear Onyx data.

The issue is that clicking back from the SSO provider's sign in page gets the user stuck on a blank screen? I think we should avoid that. In additional to the rare case of a user with multiple SSO logins, I can also see a user clicking back if they entered the wrong email address the first time.

MonilBhavsar commented:
I feel like it's very easy to get stuck or simply go to intermediate link `sign-in-with-saml/` and getting stuck with only workaround to escape is by deleting onyx data. May be we can create a followup issue to fix it. 

@NikkiWines mobile app and offline tests looks good 👍 
The only pending thing is console error https://github.com/Expensify/App/pull/28372#issuecomment-1752964037

NikkiWines committed:
fix console log for rendering and ensure users that are SAML required can get back to the login page

NikkiWines committed:
remove redundant bool cast

NikkiWines commented:
ok! fixed the console error and added some logic so that if you're SAML required and navigate to the home view (new.expensify.com/ or localhost:8082/) then it shows the login screen again!

https://github.com/Expensify/App/assets/16747822/673dc971-38c9-4429-8dba-4e3defd1a22c

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -64,23 +79,49 @@ const defaultProps = {
+    let shouldShowChooseSSOOrMagicCode = false;
+    let shouldInitiateSAMLLogin = false;
+    const platform = getPlatform();
+    if (Permissions.canUseSAML() || platform === CONST.PLATFORM.WEB || platform === CONST.PLATFORM.DESKTOP) {
```
I didn't notice in tests because is Beta's are on by default.
Should this condition be?
```suggestion
    if (Permissions.canUseSAML() && (platform === CONST.PLATFORM.WEB || platform === CONST.PLATFORM.DESKTOP)) {
```

MonilBhavsar replied:
I see beta is for other platforms, thanks for clarifying!

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -96,24 +137,33 @@ function SignInPage({credentials, account, isInModal, activeClients}) {
      * and we need it here since welcome text(`welcomeText`) also depends on it */
     const [isUsingRecoveryCode, setIsUsingRecoveryCode] = useState(false);
 
-    useEffect(() => Performance.measureTTI(), []);
-    useEffect(() => {
-        App.setLocale(Localize.getDevicePreferredLocale());
-    }, []);
```
Any idea, why this was removed?

NikkiWines replied:
🤔 i didn't intentionally remove it. Lemme see if it's on main now and if so I'll add it back in

NikkiWines replied:
it's on main, added it back!

NikkiWines committed:
re-add erroneously removed logic

situchan commented:
reviewing now

NikkiWines committed:
re-add erroneously removed logic

NikkiWines committed:
fix import

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
fix import...again

situchan commented:
Job 3 failure is from this PR. Not happening on main.
Also confirmed locally.

NikkiWines commented:
Ah, hmm probably the logic I introduced for the SAML required view so the user doesn't end up in an infinite loop 🤔

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
Revert "Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web"

This reverts commit 062fb6bcc86c8036082fa117b110750b6599b2c6.

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

NikkiWines committed:
resolve weird account state when saml required users navigate back to the login page post entering details

NikkiWines committed:
use const where appropriate

NikkiWines commented:
Ok! That should fix it - going to run through testing once more but hopefully now things work smoothly ☺️ 🙏 

https://github.com/Expensify/App/assets/16747822/50001190-a4d2-41ef-8bca-7f90cbc04c1a

As a note for myself for things to look into for a follow up: 

1. I think we might want to wrap the desktop view in something so we can add a `Go Back` button like this one so that the user can go back because right now they're _stuck_ once they enter their details and have to close and re-open the app to get away from that view. 

2. Maybe we shouldn't be calling `Navigation.navigate(ROUTES.SAML_SIGN_IN)` and instead do `return <SAMLSignInPage />` 🤔 though then we'd need to fix the intermediate loading page since it won't show anymore

NikkiWines commented:
@situchan @MonilBhavsar updated the videos for web, mweb (safari), desktop and updated the testing steps to be a bit more thorough. Will try to get updated android, mweb android, and ios videos updated as well just to be sure, but please don't wait on those to review 🙇

MonilBhavsar reviewed (commented):
Looks good and works well!
One conflict to resolve

NikkiWines commented on src/pages/signin/SignInPage.js:
```diff
@@ -64,23 +83,50 @@ const defaultProps = {
+    let shouldInitiateSAMLLogin = false;
+    const platform = getPlatform();
+    if (Permissions.canUseSAML() || platform === CONST.PLATFORM.WEB || platform === CONST.PLATFORM.DESKTOP) {
+        // True if the user has SAML required and we're not already loading their account
+        shouldInitiateSAMLLogin = hasAccount && hasLogin && isSAMLRequired && !hasInitiatedSAMLLogin && account.isLoading;
```
Oops, that comment is out of date I think - let me double check 🤔

NikkiWines replied:
@MonilBhavsar updated!

NikkiWines committed:
update comment

NikkiWines committed:
Merge branch 'main' of https://github.com/Expensify/App into nikki-saml-newdot-web

MonilBhavsar reviewed (dismissed):
Looks good and works fine!

situchan commented:
I am currently in review

MonilBhavsar commented on src/pages/signin/SignInPage.js:
```diff
@@ -64,23 +83,50 @@ const defaultProps = {
+    let shouldShowChooseSSOOrMagicCode = false;
+    let shouldInitiateSAMLLogin = false;
+    const platform = getPlatform();
+    if (Permissions.canUseSAML() || platform === CONST.PLATFORM.WEB || platform === CONST.PLATFORM.DESKTOP) {
```
Sorry, this went unnoticed as beta is universally true for dev. We need to pass beta props here

NikkiWines replied:
Discussed 1:1 w/ @MonilBhavsar we're gonna remove the beta code since the ios/android logic is being added [here](https://github.com/Expensify/App/pull/29526)

situchan commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -105,9 +122,4 @@ LogInWithShortLivedAuthTokenPage.propTypes = propTypes;
-        account: {key: ONYXKEYS.ACCOUNT},
-    }),
-)(LogInWithShortLivedAuthTokenPage);
+export default compose(withLocalize, withOnyx({account: {key: ONYXKEYS.ACCOUNT}}), withOnyx({session: {key: ONYXKEYS.SESSION}}))(LogInWithShortLivedAuthTokenPage);
```
As this is function component, we can use `useLocalize` hook.

Is there a problem using one `withOnyx` HOC? I remember this was fixed recently and no longer need multiple `withOnyx`

NikkiWines replied:
ah no, just outdated format i think. Updated!

NikkiWines committed:
remove beta, update some old styles

NikkiWines commented:
Updated once more!

situchan commented:
Code looks good. (lint failing)

I started testing with private domain email.

> Configure your account to have SAML enabled but not required

Where can I enable this?

NikkiWines commented:
> Where can I enable this?

In expensify.com under Settings > Domains > [Your Domain] > SAML

NikkiWines committed:
remove unused import

situchan commented:
> > Where can I enable this?
> 
> In expensify.com under Settings > Domains > [Your Domain] > SAML

I don't find Domains in Settings. Can you please share that url?

MonilBhavsar reviewed (approved):
This is open since long. So going to merge this now.
There is an issue with transitioning which will hopefully be fixed by https://github.com/Expensify/App/pull/28984. If not, we can address it in a separate PR

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

kavimuru commented:
@NikkiWines @MonilBhavsar Are QA steps internal?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6539299520) to staging by https://github.com/MonilBhavsar in version: 1.3.85-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

NikkiWines commented:
@kavimuru since `qa@expensifail.com` has a OneLogin account set up this doesn't need to be internally tested! However, if the please revert the `expensifail.com` domain to "SAML enabled" once testing is done 🙇

NikkiWines commented:
I'll also do some testing on staging just to try things out

NikkiWines commented:
The redirect back to NewExpensify isn't working for me on staging, but I _think_ it might be that we need to test this on prod 🤔 looking at the logs now

NikkiWines commented:
Going to hide these changes behind a beta so we can test on prod without it causing issues for customers

sakluger commented:
I commented on the beta PR, but I don't think a beta is necessary. Ideally we can test this in a day or so, and I can't think of many downsides to having it accessible to real customers. Worst case, they run into errors and we fix the bug.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6553786410) to production by https://github.com/francoisl in version: 1.3.85-4 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6553792471) to staging by https://github.com/MonilBhavsar in version: 1.3.86-0 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6567832856) to production by https://github.com/francoisl in version: 1.3.86-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

roryabraham commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -105,9 +121,7 @@ LogInWithShortLivedAuthTokenPage.propTypes = propTypes;
-)(LogInWithShortLivedAuthTokenPage);
+export default withOnyx({
+    account: {key: ONYXKEYS.ACCOUNT},
+    session: {key: ONYXKEYS.SESSION},
```
The `session` prop appears unused. Do you know if it's needed here?

MonilBhavsar replied:
We can remove it most probably. Seems like we removed the usage, but forgot to remove this key here. /cc @NikkiWines to be double sure

NikkiWines replied:
Ah yes, we can remove this 👍 I  can do that as part of https://github.com/Expensify/App/pull/29526 unless you'd like to make a separate PR fro it @roryabraham

roryabraham replied:
@NikkiWines if you want to do it as part of https://github.com/Expensify/App/pull/29526 that would be great!

NikkiWines replied:
[Done](https://github.com/Expensify/App/pull/29526/commits/5c8519f4483f40629ac77065be39197cf1b47186)!

alitoshmatov commented on src/pages/signin/SAMLSignInPage/index.js:
```diff
@@ -0,0 +1,66 @@
+    const {translate} = useLocalize();
+
+    useEffect(() => {
+        window.open(`${CONFIG.EXPENSIFY.SAML_URL}?email=${credentials.login}&referer=${CONFIG.EXPENSIFY.EXPENSIFY_CASH_REFERER}`, '_self');
```
`window.open` broke route history and user couldn't go back from SAML page. ref: https://github.com/Expensify/App/issues/29825
